File: src/domain/services/IEventBus.ts
import { EngineEvents, EventPayload, EventCallback } from "../../shared/events/EngineEvents";

export interface IEventBus {
  subscribe(event: EngineEvents | string, callback: EventCallback): () => void;
  subscribeAll(callback: (event: string, payload: EventPayload) => void): () => void;
  publish(event: EngineEvents | string, sender: string, data?: any): void;
  clearAllListeners(): void;
}

File: NovaSim/NovaSim.Core/Interfaces/IEventBus.cs
using System;

namespace NovaSim.Core.Interfaces
{
    public record EventPayload(DateTime Timestamp, string Sender, object? Data = null);

    public interface IEventBus
    {
        Action Subscribe(string @event, Action<EventPayload> callback);
        Action SubscribeAll(Action<string, EventPayload> callback);
        void Publish(string @event, string sender, object? data = null);
        void ClearAllListeners();
    }
}
